### PR TITLE
feat: handle iterable args

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
Added support to `ConversionManager` for handling `tuple` and `list` conversions (e.g. recursively applying conversions)

fixes: #376

### Checklist

- [x] Add integration support to `ape.contracts` ABI arg processing
- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape/managers/converters.py [hunk 1]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
         return to_checksum_address(value)
 
 
-hex_address_converter = HexAddressConverter(None, None)  # type: ignore
+hex_address_converter = HexAddressConverter(None, None, None)  # type: ignore
+
+
+class ListTupleConverter(ConverterAPI):
+    """
+    A converter that converts any tuple or list of items
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:100]:
```suggestion
    A converter that converts any tuple or list of items to a list of
```

File: src/ape/managers/converters.py [hunk 2]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
...
+
+    def is_convertible(self, value: Any) -> bool:
+        return isinstance(value, list) or isinstance(value, tuple)
+
+    def convert(self, value: Union[list, tuple]) -> Union[list, tuple]:
+        """
+        Convert the items inside the given list or tuple.
+
+        Args:
+            value (``list`` or ``tuple``): The collection to convert.
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:112]:
```suggestion
            value (Union[List, Tuple]): The collection to convert.
```

File: src/ape/managers/converters.py [hunk 3]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
...
+class ListTupleConverter(ConverterAPI):
+    """
+    A converter that converts any tuple or list of items
+    :class:`~ape.types.AddressType`.
+    """
+
+    def is_convertible(self, value: Any) -> bool:
+        return isinstance(value, list) or isinstance(value, tuple)
+
+    def convert(self, value: Union[list, tuple]) -> Union[list, tuple]:
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:107]:
```suggestion
    def convert(self, value: Union[lList, Tuple]) -> Union[List, Tuple]:
```

Don't these have to be the `typing` classes?

File: src/ape/managers/converters.py [hunk 4]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
...
+        """
+        Convert the items inside the given list or tuple.
+
+        Args:
+            value (``list`` or ``tuple``): The collection to convert.
+
+        Returns:
+            ``list`` or ``tuple`` (depending on input)
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:115]: Union[list, tuple]: Depending on the input

File: src/ape/managers/converters.py [hunk 5]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
...
+
+
+class ListTupleConverter(ConverterAPI):
+    """
+    A converter that converts any tuple or list of items
+    :class:`~ape.types.AddressType`.
+    """
+
+    def is_convertible(self, value: Any) -> bool:
+        return isinstance(value, list) or isinstance(value, tuple)
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:105]:
```suggestion
        return isinstance(value, (list, tuple))
```

File: src/ape/managers/converters.py (hunk 1 above)
Comment (fubuloubu) [reply src/ape/managers/converters.py:100]: Hmm, this statement is just way off entirely, isn't it?

File: src/ape/managers/converters.py [hunk 6]
```diff
@@ -119,14 +180,18 @@ def __repr__(self):
 
     @cached_property
     def _converters(self) -> Dict[Type, List[ConverterAPI]]:
+        list_tuple_converter.converter = self  # Splice this in (no need for the rest)
         converters: Dict[Type, List[ConverterAPI]] = {
             AddressType: [address_api_converter, hex_address_converter],
             bytes: [hex_converter],
             int: [],
+            Decimal: [],
+            list: [list_tuple_converter],
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:189]: Should these also be `List` and `Tuple` (from `typing` module)?

File: src/ape/managers/converters.py [hunk 7]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
...
+                    raise ConversionError(f"Unsupported ABI Type: {v.__class__}")
+
+            # Try all of them to see if one converts it over (only use first one)
+            for check_fn, convert_fn in map(lambda c: (c.is_convertible, c.convert), converters):
+                if check_fn(v):
+                    converted_value.append(convert_fn(v))
+                    conversion_found = True
+                    break
+
+                if conversion_found:
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:145]: Does this ever get execute?
I would think the break from line 143 would always happen first

File: src/ape/managers/converters.py (hunk 6 above)
Comment (fubuloubu) [reply src/ape/managers/converters.py:189]: these are concrete types, e.g. `convert(v, list)`

Tried to think of what `convert(v, List[int, str, ..., bytes])` might look like, but it seemed bad UX

File: src/ape/managers/converters.py (hunk 7 above)
Comment (fubuloubu) [reply src/ape/managers/converters.py:145]: Oh whoops, leftover from when I flattened a `for` loop

File: src/ape/managers/converters.py [hunk 8]
```diff
@@ -91,7 +92,67 @@ def convert(self, value: str) -> AddressType:
...
+                    break
```
Comment (fubuloubu) [inline src/ape/managers/converters.py:146]:
```suggestion
                    break
```

File: src/ape/managers/converters.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
-from typing import Any, Dict, List, Type
+from decimal import Decimal
+from typing import Any, Dict, List, Optional, Type, Union
```
Comment (fubuloubu) [inline src/ape/managers/converters.py:2]:
```suggestion
from typing import Any, Dict, List, Optional, Tuple, Type, Union
```

File: src/ape/managers/converters.py (hunk 6 above)
Comment (unparalleled-js) [reply src/ape/managers/converters.py:187]: Yeah that makes sense.

**Review by unparalleled-js — CHANGES_REQUESTED** [review]
This PR does not actually fix the bug it says!
lists and tuples already work, it is the fact they are passing `ContractInstance` classes as "addresses" that is causing their bug (the bug ticket wording should be changed to reflect this - I will add a comment).

We need to support `ContractInstance` in the conversion and that should resolve their bug.

**Comment by fubuloubu** [comment]
It works now!
```py
In [1]: a = accounts.load("my-alias")

In [2]: convert([a], list)
Out[2]: ['0xe868a852342Be8a43C4Ce8fC4642aaA7b857Bb82']
```

**Comment by unparalleled-js** [comment]
> It works now!
> 
> ```python
> In [1]: a = accounts.load("my-alias")
> 
> In [2]: convert([a], list)
> Out[2]: ['0xe868a852342Be8a43C4Ce8fC4642aaA7b857Bb82']
> ```

I am not still not able to pass in a `ContractInstance`

**Comment by fubuloubu** [comment]
> Deploy a smart contract containing a method that accepts an array of addresses

Sorry, this still doesn't work yet :sweat_smile: 

Still working on ABI encoding step where `args = convert(args, tuple)` will be used for `*args`

**Comment by unparalleled-js** [comment]
> > Deploy a smart contract containing a method that accepts an array of addresses
> 
> Sorry, this still doesn't work yet 😅
> 
> Still working on ABI encoding step where `args = convert(args, tuple)` will be used for `*args`
> 
> Edit: prior, the list conversion wasn't even working, e.g. this test: https://github.com/ApeWorX/ape/pull/377/files#diff-265642a4b634858b0c1bf2c7325eda296e978fa772db3b8f648365bc571ecf6fR45

Oh ok no worries! I will check out this list conversion fix though, brb

**Review by unparalleled-js — APPROVED** [review]

**Comment by fubuloubu** [comment]
@unparalleled-js 4cba949 should contain the fix, although I need to add a test for it

**Comment by fubuloubu** [comment]
Also realizing we'd probably want support for converting Transaction `kwargs` as well

**Comment by unparalleled-js** [comment]
The Project Manager needs to be updated, I am getting this error when trying to deploy:

```
    374             raise AttributeError(f"{self.__class__.__name__} has no attribute '{attr_name}'.")
    375 
--> 376         return ContractContainer(  # type: ignore
    377             contract_type=contract_type, _provider=self.networks.active_provider
    378         )

TypeError: __init__() missing 1 required positional argument: '_converter'
```

**Comment by fubuloubu** [comment]
> The Project Manager needs to be updated, I am getting this error when trying to deploy:
> 
> ```
>     374             raise AttributeError(f"{self.__class__.__name__} has no attribute '{attr_name}'.")
>     375 
> --> 376         return ContractContainer(  # type: ignore
>     377             contract_type=contract_type, _provider=self.networks.active_provider
>     378         )
> 
> TypeError: __init__() missing 1 required positional argument: '_converter'
> ```

fixed in 873263d

**Comment by unparalleled-js** [comment]
Okay, so now I can't call my method that takes a list anymore. I always get the `ArgumentsLengthError()`

**Comment by unparalleled-js** [comment]
Fo

> Okay, so now I can't call my method that takes a list anymore. I always get the `ArgumentsLengthError()`
> 
> ArgumentsLengthError: The number of the given arguments do not match what is defined in the ABI.

Setting breaking points, it seems to be doubling my args. Like if I give it a list with two addresses in it (one an account and one a contract) it will turn it into two lists in `ContractTransactionHandler` line 173.

```
In [8]: c.fund([c, a], value=100, sender=a2)
> /ape/src/ape/contracts/base.py(174)__call__()
-> selected_abi = _select_abi(self.abis, args)
(Pdb) args
self = fund(address[] test)
args = ([<Fund 0x274b028b03A250cA03644E6c578D81f019eE1323>, <TestAccount 0x1e59ce931B4CFea3fe4B875411e280e173cB7A9C>],)
(Pdb) c
> /ape/src/ape/contracts/base.py(176)__call__()
-> raise ArgumentsLengthError()
(Pdb) args
self = fund(address[] test)
args = (['0x274b028b03A250cA03644E6c578D81f019eE1323', '0x1e59ce931B4CFea3fe4B875411e280e173cB7A9C'], ['0x274b028b03A250cA03644E6c578D81f019eE1323', '0x1e59ce931B4CFea3fe4B875411e280e173cB7A9C'])
(Pdb) 
```

File: src/ape/managers/converters.py [hunk 10]
```diff
@@ -91,7 +92,52 @@ def convert(self, value: str) -> AddressType:
...
+        """
+
+        converted_value: List[Any] = []
+
+        for v in value:
+            # Try all of them to see if one converts it over (only use first one)
+            conversion_found = False
+            # NOTE: Double loop required because we might not know the exact type of the inner
+            #       items. The UX of having to specify all inner items seemed poor as well.
+            for typ in self.converter._converters:
```
Comment (unparalleled-js) [inline src/ape/managers/converters.py:124]: I think we may need to filter out `list` and `tuple` here - that could explain why my lists are doubling.

File: src/ape/contracts/base.py [hunk 11]
```diff
@@ -18,11 +18,6 @@
     from ape.managers.networks import NetworkManager
 
 
-def _encode_address_args(*args):
-    # Convert higher level address types to str
-    return [arg.address if isinstance(arg, AddressAPI) else arg for arg in args]
```
Comment (unparalleled-js) [inline src/ape/contracts/base.py:23]: I am guessing you are going to add a converter that does this?

File: src/ape/managers/converters.py (hunk 10 above)
Comment (fubuloubu) [reply src/ape/managers/converters.py:124]: naw, fixed this in c7ddce6

File: src/ape/contracts/base.py (hunk 11 above)
Comment (fubuloubu) [reply src/ape/contracts/base.py:23]: we already have one!

Comment (fubuloubu) [reply src/ape/contracts/base.py:23]: This was there as a stopgap because we didn't do the full conversion step inside of `ape/contracts` methods yet

Comment (fubuloubu) [reply src/ape/contracts/base.py:23]: Next would come the elimination of `_encode_address_kwargs` as well

Comment (fubuloubu) [reply src/ape/contracts/base.py:23]: `_encode_address_kwargs` removed in 875cd73

**Review by unparalleled-js — APPROVED** [review]
It works!

**Pull request merged**